Indexes matches

lukescott opened the pull request:
Lets say you have a file structure like this:

* `section1/main.js`
* `section2/main.js`

And you write this:

```js
import * as sections from './*/main.js'
```
Your identifiers end up being `section1$main` and `section2$main`. If the folder ends up being the relevant portion, your identifiers get very wordy. But, depending on how complex your pattern is, it's difficult to determine how to figure that out automatically.

So I added support for this:

```js
import { $0 as sections } from './*/main.js'
```

And now my identifiers are `section1` and `section2`.

I break apart the glob pattern and specify which part of the pattern I want to use. If I specify the filename part it strips extension as expected.

NOTE: I branched from #37

novemberborn commented:
> ```
> import { $0 as sections } from './*/main.js'
> ```
> And now my identifiers are `section1` and `section2`.

What is the significance of `$0 as sections`? What would happen if I imported `$0 as foo` instead?

novemberborn commented:
I took the opportunity to stop targeting unsupported Node.js versions, and remove the Babel build pipeline, see https://github.com/novemberborn/babel-plugin-import-glob/pull/40. Unfortunately it means there's now a heck of a merge conflict here, sorry 😢

lukescott commented:
`{$0 as sections}` works just like `* as sections`. The difference is you specify which part of the glob pattern I want to grab. It's sort of like when you use `()` in a RegEx and you want to grab a specific match. A glob pattern is broken up into sub-patterns using `/`. So I ignore the literal parts and index the dynamic parts.

So if you had `./*/foo/*.handlebars.js` `$0` would refer to the first `*`. `$1` would refer to `*.handlebars.js`. With `* as templates` you would end up with `folder$foo$template` as your identifier. Using `$0` it would be `folder`. With `$1` it would be `template`.

I'll just re-apply your changes since I re-wrote a large portion of the code.

lukescott commented:
@novemberborn I addressed the merge conflicts, added more tests and made sure the branch coverage was at 100%.

I also changed it so if you don't prefix with `glob:`, it no longer assumes relative if `.` is not included. It only does so with `glob:`, to maintain backwards compatibility. I did this because without the prefix, it would have looked like you were trying to include from a package. Since that is an undesirable assumption, I made it more strict.

novemberborn commented:
I see where you're going with this now. I'm not convinced by the syntax though, and there seems to be a significant increase in complexity to make it all work.

I think in the end this plugin is intended for cases where writing individual import statements is overly cumbersome. Globbing can help there, but it has its own quirks and limitations. This is the trade-off you're making when dropping in a Babel plugin. Attempting to smooth over these rough edges in the import syntax seems futile; it'd be better to accept them and if necessary transform the imported objects in code.

I appreciate the effort you've been putting into this, but unfortunately I won't be accepting this PR.

---

> I also changed it so if you don't prefix with glob:, it no longer assumes relative if . is not included. It only does so with glob:, to maintain backwards compatibility. I did this because without the prefix, it would have looked like you were trying to include from a package. Since that is an undesirable assumption, I made it more strict.

Good thinking. I'll be releasing a major anyway so I'm not too concerned about backwards compatibility. I think with `glob:` too we should require it to start with `./` or `../`. Would you want to open a PR for that?

lukescott commented:
With my specific use case, I'm importing a bunch of plugin-based Redux reducers:

```js
import {$0 as reducerMap} from "./*/reducer.js"

function reducer(state, action) {
    // simplified version of root reducer
    return reducerMap[state.key](state, action);
}
```

Without the above I would have to do something like this:
```
import * as reducerMap from "./*/reducer.js"

function reducer(state, action) {
    // simplified version of root reducer
    return reducerMap[state.key+'$reducer'](state, action);
}
```

For me my folder structure isn't as bad, but it could be worse. I could re-map it in code:

```
const reducers = _.mapValues(reducerMap, v => v.split('$')[0])
```

But in both cases I'm doing runtime resolution. First one I'm doing a concat on each execution, and the second I'm doing it once at the start.

The point of the plugin is to do all this at compile time, hence the `{$0 as reducerMap}` syntax. I wasn't able to come up with a better syntax unfortuantly. `$0 as reducerMap` (without the `()`) is invalid syntax. I needed a way to select a particular match. I could have added `#0` to the end of the file name, but that felt worse.

In a lot of areas I would say the code is less complex - I pulled out all the helper functions out of the main function body. The only real addition is the pattern-splitting, which is based on makeRe function in minimatch (lib used by glob).

If you don't like the syntax and don't care about backwards compatibility, we could do this instead:

1 - Remove `common-path-prefix` and default to the first dynamic portion. So with `./*/reducer.js` it would be the first star seen in the pattern. With `templates/*.handlebars.js` it would be `*.handlebars.js`, reduced to `*` by `common-extname`. -- So `* as reducers` would be the same as `{$0 as reducers}`.

2 - Put the `$[0-9]` behind a plugin config option (off by default). Only point to this is if you wanted `$1` or larger.

If we did just the first (note keep `$[0-9]` in some way) my current use case would be solved.

novemberborn commented:
> 1 - Remove common-path-prefix and default to the first dynamic portion. So with ./*/reducer.js it would be the first star seen in the pattern.

How would that influence the identifier? Wouldn't it still add a `$reducer` suffix?

What's interesting about your use case is that the directory name is unique, but the filename isn't. Perhaps we should remove trailing, fixed path segments:

* `./*/reducer.js` would generate identifiers based on the matched directory names, stripping the fixed trailing `reducer.js` segment
* `./templates/*.handlebars.js` would strip the common `templates` prefix, and the common `.handlebars.js` extname

lukescott commented:
> > 1 - Remove common-path-prefix and default to the first dynamic portion. So with ./*/reducer.js it would be the first star seen in the pattern.

> How would that influence the identifier? Wouldn't it still add a $reducer suffix?

It would not. It would just use the `*` portion.

Multimatch (glob's matching lib) breaks up the pattern by `/`. So `./*/reducer.js` becomes:

- `.` (string)
- `*` (RegExp)
- `reducer.js` (string)

This is found on `gm.multimatch.set` (gm is returned by GlobSync - gm.found is the same array glob.sync() returns).

There is a function on the multimatch library called makeRe which combines those parts into a single RegExp. I basically do the same exact thing, but I wrap the RegExp parts of the expression with `()`'s. That makes each "dynamic portion" indexed in "...".match()'s result.

I'm not comparing each path matched for commonality, but rather taking the part of the path that I know will be different. So `./*/reducer.js` would not include `$reducer` because I'm pulling it from the matched result directly.

You could change `common-extname` to `common-path-suffix` to eliminate `$reducer` as well, but it would be more efficient to pull it out of the match.

Where the result would differ is if you had a path like `./*/*.reducer.js`. If you assume the first index, that would also remove `$[star]` (assuming `common-extname` removed the `.reducer.js` part). Having two `*` probably wouldn't be common though - but if it were to happen, how would we handle it? There are a few ways:

- Join the min to max indexes with `$`
- Let the user choose with a `${number}` selector
- Both of the above
- Assume last `*` instead of first, so `*.reducer.js`
- More than one `*` is an error

lukescott commented:
Ok, so after thinking about it, I'm just going to close this and provide an alternative and explain that, and hopefully that makes more sense.
